website/blog/2019-07-22-presto-unlimited-mpp-database-at-scale.md

```diff
@@ -0,0 +1,123 @@
+---
+title: Presto Unlimited: Open-Source MPP Database at Scale
+author: Wenlei Xie
+authorURL: https://www.linkedin.com/in/wenleix/
+authorFBID: 681470066
+---
+
+Wenlei Xie, Andrii Rosa, Shixuan Fan, Rebecca Schlussel, Tim Meehan
+
+## Introduction 
+
+Presto is an open source distributed SQL query engine for running analytic queries against data sources of all sizes ranging from gigabytes to petabytes.
+
+Presto was originally designed for interactive use cases, however, after seeing the merit in having a single interface for both batch and interactive, it is now also used heavily for processing batch workloads. As a concrete example, more than 80% of new warehouse batch workloads at Facebook are developed on .Presto. Its flexible “connector” design makes it possible to run queries against heterogeneous data sources — such as joining together Hive and MySQL tables without preloading the data.
+
+However, memory-intensive (many TBs) and long-running (mulitple hours) queries have been major pain points for Presto users. It is difficult to reason how much memory queries will use and when it will hit memory limit, and failures in long-running queries cause landing time variance. To improve user experience and scale MPP Database to large ETL workloads, we started this Presto Unlimited project.
+
+## Grouped Execution
+
+[Grouped execution](https://github.com/prestodb/presto/pull/8951) was developed to scale Presto to run memory-intensive queries by leveraging table partitioning. 
+
+Consider the following query, where table `customer` and `orders` are already bucketed on `custkey`: 
+
+```
+SELECT ...
+FROM customer JOIN orders
+    USING custkey
+```
+
+Without grouped execution, the workers will build hash table using all the data on the build side (Table `orders`):
+
+![Ungrouped Execution](/img/blog/2019-07-22-presto-unlimited-mpp-database-at-scale/ungrouped.png)
+
+However, since `customer` and `orders` are already bucketed, Presto can schedule the query execution in a more intelligent way to reduce peak memory consumption: for each bucket `i`, joining the bucket `i` on table `customer` and `orders` can be done independently! In Presto engine, we call this computation unit a “lifespan”:
+
+![Grouped Execution](/img/blog/2019-07-22-presto-unlimited-mpp-database-at-scale/grouped.png)
+
+Grouped execution has been enabled in Facebook production workload for over a year, supports queries that would otherwise require tens of TBs, in some cases over 100 TB, of distributed memory.
+
+## Presto Unlimited 
+
+In this section we are going to introduce the two projects we worked on in the last half, aims at the two pain points discussed in the introduction section:
+
+- Exchange materialization for memory-intensive queries (many TBs)  
+- Recovery grouped execution for long-running queries (multiple hours)
+
+A few initial production pipelines in Facebook warehouse is already empowered by these two features. 
```

Comment:
Rigorously speaking, we haven't turned on recoverable grouped execution in production yet. Current release might be rolled back so I'm not confident about when this would happen.